src/Razor/src/Microsoft.VisualStudio.LanguageServerClient.Razor/HtmlCSharp/DefaultLSPProgressListener.cs

```diff
@@ -0,0 +1,136 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Concurrent;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.VisualStudio.LanguageServer.Client;
+using Microsoft.VisualStudio.LanguageServer.Protocol;
+using Newtonsoft.Json.Linq;
+using System.Composition;
+
+namespace Microsoft.VisualStudio.LanguageServerClient.Razor.HtmlCSharp
+{
+    [Shared]
+    [Export(typeof(LSPProgressListener))]
+    internal class DefaultLSPProgressListener : LSPProgressListener, IDisposable
+    {
+        private readonly ILanguageServiceBroker2 _languageServiceBroker;
+
+        [ImportingConstructor]
+        public DefaultLSPProgressListener(ILanguageServiceBroker2 languageServiceBroker)
+        {
+            if (languageServiceBroker is null)
+            {
+                throw new ArgumentNullException(nameof(languageServiceBroker));
+            }
+
+            _languageServiceBroker = languageServiceBroker;
+
+            _languageServiceBroker.ClientNotifyAsync += ClientNotifyAsyncListenerAsync;
+        }
+
+        private ConcurrentDictionary<string, CallbackRequest> ActiveRequests { get; }
+            = new ConcurrentDictionary<string, CallbackRequest>();
+
+        public override async Task ClientNotifyAsyncListenerAsync(object sender, LanguageClientNotifyEventArgs args)
+        {
+            await ProcessProgressNotificationAsync(args.MethodName, args.ParameterToken).ConfigureAwait(false);
+        }
+
+        private async Task ProcessProgressNotificationAsync(string methodName, JToken parameterToken)
+        {
+            if (methodName != Methods.ProgressNotificationName ||
+               !parameterToken.HasValues ||
+               parameterToken["value"] is null ||
```

Comment:
Does this throw if there's no "value" key? Asking because given this can be called as a result of any IProgress<T> notification I'm curious if we should be more defensive and do tryget's

Reply:
Undefined keys are simply `null` (no key not found exception) so this shouldn't be an issue 😄